clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -4692,6 +4692,17 @@ bool Sema::InstantiateDefaultArgument(SourceLocation CallLoc, FunctionDecl *FD,
                                       ParmVarDecl *Param) {
   assert(Param->hasUninstantiatedDefaultArg());
 
+  // FIXME: We don't track member specialization info for non-defining
+  // friend declarations, so we will not be able to later find the function
+  // pattern. As a workaround, don't instantiate the default argument in this
+  // case. This is correct per wording and only an error recovery issue, as per
```

Comment:
```suggestion
  // case. This is correct per the standard and only an issue for recovery purposes.
```

Reply:
I'm not even sure the fixme is worth it here. We already produce an error, another one would not be particularly useful or worth the refactoring cost

@erichkeane

Reply:
Substitution failures in the default argument are no longer diagnosed when the friend function declaration has no definition. Since this was already an error, I think it’s now just worse error recovery. Or should the checker proceed with instantiation and error recovery?

Reply:
We attempt to try to recover whenever possible, as that results in a more 'correct' AST for tooling and for us evaluating it later.  That said, we do have `RecoveryExpr` for cases where we can't really do anything with it, but it needs a bit of testing for every new 'thing', because it is always dependent and that causes consternation with some sections of code.

Reply:
to move forward with this PR, should I add additional tests for recovery scenarios, or are there specific changes needed to the recovery logic?

Reply:
I would be sufficiently satisfied with the tests being here, perhaps with a 'FIXME' in the test if we come up short.

Reply:
@erichkeane okay, thanks. As far as I remember, this case isn't processed with recovery; however, I'll check it anyway.

Reply:
I messed with this for a little bit and can't come up with any good way to hit this with a recoveryexpr/etc, so I think I'm probably ok leaving this as a 'fixme'.  I think we could/do better here someday if we can come up with an example, but nothing I can conceive now.

@a-tarasyuk : feel free to merge this now, and revisit if/when you think of a good repro here that we can examine/improve.

Reply:
The tests for this error recovery issue is already there, it was suggested and included in the original PR.
See test case marked with FIXME.